File: ui-react/src/api/license.ts
// Offline license API — thin typed wrappers over the Tauri `license_*` commands,
// which delegate to the shared `platypus-license` Ed25519 verifier. Mirrors the
// `LicenseInfo` shape the Rust command emits (serde camelCase).
import { invoke } from "@tauri-apps/api/core";
import { useCallback, useEffect, useState } from "react";

export type LicenseStatus =
  | "valid"
  | "expired"
  | "not_yet_valid"
  | "machine_mismatch"
  | "bad_signature"
  | "malformed"
  | "missing";

export interface LicenseInfo {
  status: LicenseStatus;
  valid: boolean;
  id: string | null;
  name: string | null;
  email: string | null;
  plan: string | null;
  tier: string | null;
  seats: number | null;
  features: string[];
  /** Expiry, unix seconds; null = perpetual. */
  expires: number | null;
  /** Bound machine fingerprint; null = floating. */
  machine: string | null;
}

const isTauri = (): boolean =>
  typeof window !== "undefined" &&
  ("__TAURI_INTERNALS__" in window || "__TAURI__" in window);

/** What the UI shows before/without a Tauri backend (e.g. the web build). */
export const UNLICENSED: LicenseInfo = {
  status: "missing",
  valid: false,
  id: null,
  name: null,
  email: null,
  plan: null,
  tier: null,
  seats: null,
  features: [],
  expires: null,
  machine: null,
};

export const licenseApi = {
  /** Current status from disk. Never throws; returns `missing` off-Tauri. */
  async status(): Promise<LicenseInfo> {
    if (!isTauri()) return UNLICENSED;
    return invoke<LicenseInfo>("license_status");
  },
  /** Verify + persist a token. Rejects (throws the reason) when not valid. */
  async activate(token: string): Promise<LicenseInfo> {
    return invoke<LicenseInfo>("license_activate", { token });
  },
  /** Remove the stored license. */
  async deactivate(): Promise<void> {
    if (!isTauri()) return;
    await invoke("license_deactivate");
  },
  /** This machine's node-lock fingerprint, to include in a license request. */
  async machineFingerprint(): Promise<string | null> {
    if (!isTauri()) return null;
    return invoke<string | null>("machine_fingerprint");
  },
};

/** Client-side entitlement check (the backend re-checks for paid commands). */
export function hasFeature(info: LicenseInfo | null, feature: string): boolean {
  if (!info || !info.valid) return false;
  return info.features.includes("*") || info.features.includes(feature);
}

/** Subscribe a component to the current license status. */
export function useLicense() {
  const [info, setInfo] = useState<LicenseInfo | null>(null);
  const [loading, setLoading] = useState(true);

  const refresh = useCallback(async () => {
    setLoading(true);
    try {
      setInfo(await licenseApi.status());
    } finally {
      setLoading(false);
    }
  }, []);

  useEffect(() => {
    void refresh();
  }, [refresh]);

  return { info, loading, refresh };
}

File: experimental/ui-react/src-tauri/src/license.rs
//! Tauri command surface for offline, node-locked license verification.
//!
//! The verifier itself lives in the `platypus-license` crate (shared with the
//! `platypus` Python module); this module only adds the desktop concerns:
//! where the token is stored, how it's activated from the UI, and the backend
//! feature gate.
//!
//! Storage: the activated token is written verbatim to `<cache>/license.plt`
//! (the same `cache_dir` the project uses). It's a signed, tamper-evident blob,
//! so plain-text-on-disk is fine — editing it just makes it fail verification.

use std::path::PathBuf;

use serde::Serialize;
use tauri::State;

use platypus_license as lic;

use crate::state::AppState;

const LICENSE_FILE: &str = "license.plt";

fn license_path(state: &AppState) -> PathBuf {
    state.cache_dir.join(LICENSE_FILE)
}

/// The stored token, trimmed; `None` if absent or blank.
fn read_token(state: &AppState) -> Option<String> {
    std::fs::read_to_string(license_path(state))
        .ok()
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
}

/// Flattened license view for the frontend (camelCase to match the TS layer).
/// Claim fields are `null` when the token didn't verify.
#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct LicenseInfo {
    pub status: String,
    pub valid: bool,
    pub id: Option<String>,
    pub name: Option<String>,
    pub email: Option<String>,
    pub plan: Option<String>,
    pub tier: Option<String>,
    pub seats: Option<u32>,
    pub features: Vec<String>,
    pub expires: Option<i64>,
    pub machine: Option<String>,
}

impl From<lic::Verified> for LicenseInfo {
    fn from(v: lic::Verified) -> Self {
        let c = v.claims;
        LicenseInfo {
            status: v.status.as_str().to_string(),
            valid: v.status.is_valid(),
            id: c.as_ref().map(|c| c.id.clone()),
            name: c.as_ref().map(|c| c.name.clone()),
            email: c.as_ref().map(|c| c.email.clone()),
            plan: c.as_ref().map(|c| c.plan.clone()),
            tier: c.as_ref().map(|c| c.tier.clone()),
            seats: c.as_ref().map(|c| c.seats),
            features: c.as_ref().map(|c| c.features.clone()).unwrap_or_default(),
            expires: c.as_ref().and_then(|c| c.expires),
            machine: c.as_ref().and_then(|c| c.machine.clone()),
        }
    }
}

/// Human-readable reason an activation was refused.
fn activation_error(status: lic::Status) -> String {
    match status {
        lic::Status::Expired => "This license has expired.".into(),
        lic::Status::NotYetValid => "This license is not valid yet (check your clock).".into(),
        lic::Status::MachineMismatch => {
            "This license is locked to a different machine.".into()
        }
        lic::Status::BadSignature => "Invalid license: signature check failed.".into(),
        lic::Status::Malformed => "That doesn't look like a Platypus license key.".into(),
        lic::Status::Missing => "No license key was provided.".into(),
        lic::Status::Valid => "License is valid.".into(),
    }
}

// ── Commands ─────────────────────────────────────────────────────────────────

/// Current license status, read from disk. Never errors — an absent/blank file
/// is reported as `missing`.
#[tauri::command]
pub fn license_status(state: State<'_, AppState>) -> LicenseInfo {
    lic::evaluate_now(read_token(&state).as_deref()).into()
}

/// This machine's node-lock fingerprint (32 hex chars), to paste into a license
/// request. Matches the Python `platypus.license.machine_fingerprint()` output.
#[tauri::command]
pub fn machine_fingerprint() -> Option<String> {
    lic::fingerprint::machine_fingerprint()
}

/// Verify `token`; on a `valid` outcome persist it and return the info. Any
/// non-valid outcome writes nothing and returns the reason as an error.
#[tauri::command]
pub fn license_activate(
    token: String,
    state: State<'_, AppState>,
) -> Result<LicenseInfo, String> {
    let verified = lic::evaluate_now(Some(&token));
    if !verified.status.is_valid() {
        return Err(activation_error(verified.status));
    }
    std::fs::write(license_path(&state), token.trim())
        .map_err(|e| format!("could not save license: {e}"))?;
    Ok(verified.into())
}

/// Remove the stored license (revert to unlicensed). Idempotent.
#[tauri::command]
pub fn license_deactivate(state: State<'_, AppState>) -> Result<(), String> {
    match std::fs::remove_file(license_path(&state)) {
        Ok(()) => Ok(()),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(e) => Err(format!("could not remove license: {e}")),
    }
}

// ── Backend feature gate ─────────────────────────────────────────────────────

/// Reject a command unless the stored license is `valid` *and* grants `feature`.
/// Call at the top of any paywalled command, e.g.:
///
/// ```ignore
/// crate::license::require_feature(&state, "taint")?;
/// ```
///
/// Enforcement is **release-only**: debug builds (`cargo run`/`tauri dev`) skip
/// the check so day-to-day development isn't gated, while shipped binaries
/// (`cargo build --release`) require a license. This is defense-in-depth behind
/// the frontend gate, not the primary UX.
pub fn require_feature(state: &AppState, feature: &str) -> Result<(), String> {
    if cfg!(debug_assertions) {
        return Ok(());
    }
    let verified = lic::evaluate_now(read_token(state).as_deref());
    if !verified.status.is_valid() {
        return Err(format!(
            "A valid license is required for this feature ({}). Add one in Settings → License.",
            verified.status.as_str()
        ));
    }
    if !verified.claims.as_ref().is_some_and(|c| c.has_feature(feature)) {
        return Err(format!("Your license does not include the '{feature}' feature."));
    }
    Ok(())
}
